# AESPA: Accurate and Efficient Sub-sampling Pipeline for WGS analysis

## Overview
This pipeline is designed to perform quality control (QC) for whole genome sequencing (WGS) data by conducting 3x subsampling.
![flowchart](./Figure/flow_chart.png)
The entire workflow is modularized using Nextflow.

## üåü Features
Efficient QC Processing: Subsampling and QC for WGS data.
Modular Design: All processes are modularized for ease of maintenance and customization.
Conda Integration: Each sub-workflow has a dedicated Conda environment.

## üìã Requirements
Nextflow: Version 23.10.1 or higher
Sun Grid Engine: For job scheduling
Conda: For managing software dependencies

## üìÇ Directory Structure

```arduino
wgs_qc
‚îú‚îÄ‚îÄ apps
‚îÇ   ‚îî‚îÄ‚îÄ GenomeAnalysisTK-3.7
‚îú‚îÄ‚îÄ bin
‚îÇ   ‚îú‚îÄ‚îÄ batch.sh
‚îÇ   ‚îú‚îÄ‚îÄ DOC_distance.py
‚îÇ   ‚îú‚îÄ‚îÄ extract_variants
‚îÇ   ‚îú‚îÄ‚îÄ samtools_flagstat.py
‚îÇ   ‚îú‚îÄ‚îÄ sqs_generate.py
‚îÇ   ‚îú‚îÄ‚îÄ stat_summary.py
‚îÇ   ‚îú‚îÄ‚îÄ subsampler.sh
‚îÇ   ‚îú‚îÄ‚îÄ subsampling.py
‚îÇ   ‚îî‚îÄ‚îÄ summary_stat.py
‚îú‚îÄ‚îÄ conf
‚îÇ   ‚îú‚îÄ‚îÄ reference.json
‚îÇ   ‚îú‚îÄ‚îÄ sge_conda.config
‚îÇ   ‚îî‚îÄ‚îÄ sge_local.config
‚îú‚îÄ‚îÄ input.json
‚îú‚îÄ‚îÄ main.nf
‚îú‚îÄ‚îÄ nextflow.config
‚îú‚îÄ‚îÄ run.nf_test.sh
‚îú‚îÄ‚îÄ src
‚îÇ   ‚îú‚îÄ‚îÄ genome.dict
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.amb
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.ann
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.bwt
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.fai
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.pac
‚îÇ   ‚îú‚îÄ‚îÄ genome.fa.sa
‚îÇ   ‚îú‚îÄ‚îÄ genome.gff
‚îÇ   ‚îú‚îÄ‚îÄ genome.gtf
‚îÇ   ‚îî‚îÄ‚îÄ sorted-reference.xml
‚îî‚îÄ‚îÄ workflow
    ‚îú‚îÄ‚îÄ bam_stat_calculation.nf
    ‚îú‚îÄ‚îÄ bam_stat_calculation.picard.yml
    ‚îú‚îÄ‚îÄ bam_stat_calculation.yml
    ‚îú‚îÄ‚îÄ iSAAC_pipeline.nf
    ‚îú‚îÄ‚îÄ iSAAC_pipeline.yml
    ‚îú‚îÄ‚îÄ preprocessing.nf
    ‚îú‚îÄ‚îÄ preprocessing.yml
    ‚îú‚îÄ‚îÄ strelka_variant_call.nf
    ‚îú‚îÄ‚îÄ strelka_variant_call.yml
    ‚îú‚îÄ‚îÄ summary_qc_stat.nf
    ‚îî‚îÄ‚îÄ summary_qc_stat.yml
```

## üõ† Installation

```sh
cd workflow/
conda env create -f workflow/preprocessing.yml
conda env create -f workflow/iSAAC_pipeline.yml
conda env create -f workflow/bam_stat_calculation.yml
conda env create -f workflow/strelka_variant_call.yml
conda env create -f workflow/summary_qc_stat.yml
```

## üöÄ Usage
### Running the Pipeline
To run the pipeline with Conda environments enabled:
```sh
nextflow run main.nf -profile sge_conda_env --json_file input.json
```

To run the pipeline with local environments:
```sh
nextflow run main.nf -profile sge_local_env --json_file input.json
```

### Example input.json
```json
{
    "raw_forward_input": "/your/input_R1.fastq.gz",
    "raw_reverse_input": "/your/input_R2.fastq.gz",
    "sample_sheet": "/path/your/SampleSheet.csv",
    "sample_id": "test",
    "result_dir": "/path/your/output"
}
```

### Example Script

```sh
#!/bin/bash
outdir=test
work=$outdir/work
json_file=input.json
nf_temp=$outdir/temp
export NXF_WORK=$work
export NXF_TEMP=$nf_temp
export NXF_OFFLINE=true # if it is not connected to internet.

mkdir -p $outdir
mkdir -p $work
mkdir -p $nf_temp

nextflow run main.nf \
    -profile sge_conda_env \
    --json_file $json_file \
    -with-report "$outdir/reports/nf_out.report.html" \
    -with-dag "$outdir/reports/flowchart.png" \
    -with-timeline "$outdir/reports/nf_out.timeline.report.html" \
    --log "$outdir/nxf.log" \
    -resume \
    -with-trace "$outdir/reports/trace.txt" \
    &> $outdir/run.log.txt
```

## üß¨ Workflow Details
Submodules
Preprocessing (preprocessing.nf)
iSAAC Alignment (iSAAC_pipeline.nf)
BAM Statistics Calculation (bam_stat_calculation.nf)
Variant Calling (strelka_variant_call.nf)
QC Summary (summary_qc_stat.nf)

## Example for bam_stat_calculation.yml
```yaml
name: calc_bam_stat
channels:
  - defaults
  - conda-forge
  - bioconda
dependencies:
  - python=3.11.3
  - verifybamid2=2.0.1
  - bedtools
  - pip
  - pip:
      - scipy
      - pysam
```

## üìÑ Attached Reference Genome
RefSeq version Human Genome (GRCh38;hg38) without scaffolds.

## ‚öôÔ∏è Customization
Conda Integration: Enable Conda environments with conda.enabled=true.
Local Environment: Configure paths in conf/sge_local.config.
Change Reference: Modify the file paths on the reference json file in conf/reference.json

## üìä Performance
Speed: 3-4 times faster than the traditional iSAAC pipeline.
Resource Usage: Approx. 1 hour runtime with a max memory usage of 40GB.
Accuracy: Moderate error (MSE 3-5) with accurate mappable mean depth, deduplication rate, and contaminated reads.
